### ISSUE TITLE
Proposal for Flexible module/package versioning for Collector and Collector-contrib repos

### ISSUE THREAD

**Issue by eddyleelin** [open-telemetry/opentelemetry-collector#3449]
[Detailed design doc here.](https://docs.google.com/document/d/1d6vAvUiqU1f5Y4aOozImIgjVO_l2jB36ImfHLgzAM-A/edit?usp=sharing)

**TL;DR:** Rather than versioning an entire repo (e.g. Collector contrib) with a single shared version, **every** individual component (e.g. OTLP receiver, Prometheus exporter, etc.) should be defined as a Go Module and can thus be versioned individually. To make versioning and releasing easier, these Modules/components can then be grouped into sets of Modules. Each set of Modules can maintain a shared version (i.e. all modules in the set can be updated at the same time) easily using automated tools already created to solve the same problem in the OpenTelemetry-Go repo.

### Proposal
**The "Problem"**
The OpenTelemetry-Collector repos (core and contrib) contain many components (e.g. OTLP Receiver, OTLP Exporter, Prometheus Exporter, etc.), some of which may be considered stable and some of which may be considered unstable. Currently, the same version is shared across all components in the repo, making it hard to keep track of which stage each component is in, especially with more components continually being added.

With the [proposal to reorganize the Collector repositories into multiple repos](https://github.com/open-telemetry/opentelemetry-collector/issues/3627), a common concern is that releasing new versions of the "core" or "API" repo (on which the other repos depend on) will take even longer to perform, even now. As @jrcamp mentioned in the [proposal](https://docs.google.com/document/d/1XyYyaYtc1R8-es0BL65XE21dORLQpCPM1Pinw4MyEYs/edit), "Today doing a release can takes days and 3 repositories will just be even worse unless there's a lot of automation."

**The Solution**
To address the first problem of needing multiple modules with different versions in the same repo, different components in the Collector repo should maintain separate versions by utilizing the concept of Go Modules. If each component is encapsulated as a Go Module (defined by a go.mod file), it is easy to version each Module/component individually. However, there is no official support for versioning and releasing multiple Go Modules within a single repo, making it hard to manually version and release multiple Go Modules at the same time.

To make things easier, we can group several Modules/components into sets of Modules. Each set of Modules can maintain a shared version (i.e. all modules in the set can be updated at the same time) easily using the Releaser tool in OpenTelemetry-Go-Build-Tools ([PR here](https://github.com/open-telemetry/opentelemetry-go-build-tools/pull/8)) already created to solve the same problem in the OpenTelemetry-Go repo.

In addition, the Releaser tool makes releasing new versions easier, as it automates the majority of the releasing process using scripts written in Go. A proposal to be added is a "sync" feature, where once the versions have been updated in the core API, references in contrib/components to the core modules will be automatically updated to have the new version numbers.

**In summary, this enhancement will help...**

- Clearly distinguish each component’s stability status via the version number.
- Automate the release process for Collector-related repos, reducing the amount of time it takes to release new versions for components.

### Design
The implementation we have in mind would support the following steps:

- Create a way to declare sets of Modules that maintain consistent versions in a YAML file in each repo ([example from OpenTelemetry-Go](https://github.com/open-telemetry/opentelemetry-go/blob/main/versions.yaml)). This will help with automating the releasing process and making it very clear to users which components are stable/unstable, as well as showing a clear logical grouping between similar components.
- Using this versioning manifest file, we are able to run a Cobra binary which automates the release of Modules.
  - First a “verify” subcommand checks that the versioning manifest file is valid:
    - All packages/modules are contained in a set.
    - No more than one set of modules exists for any given non-zero major version.
    - No dependencies on any experimental module exist in stable modules.
  - Next, a “prerelease” subcommand creates a new branch, edits all go.mod files as necessary, and commits the changes.
  - Lastly, a “tag” subcommand generates Git Tags for the commit, making version numbers public.

**Comment by eddyleelin** [comment]
Here is a helpful diagram that overviews the proposed solution: 
![Collector Module Verisoning Design Diagram](https://user-images.githubusercontent.com/30187949/122258405-fe356980-ce9e-11eb-8d63-92f9ca2e0b2c.png)

**Comment by alolita** [comment]
This module-management tool could be added to the build-tools repo or in a new repo (we'd like to avoid the new repo scenario since build-tools should be easy for all maintainers to use for their tooling.)

Thoughts on where this tool should sit? Please advise @bogdandrutu @jpkrohling

**Comment by jpkrohling** [comment]
Questions:

1. I believe this to be the case, but can you please confirm that the regular Go tools will continue to be used for other purposes since go.mod is still the source of truth for them? 
1. The intermediary branch will act as tags then, right? The branch `stable-v1` will be used by downstream dependencies of the collector when they resolve the transitive dependencies. Like:

```
my-component -> groupbytrace v0.30.0 -> groupbysignal stable-v1
```

Is that the case?

**Comment by alolita** [comment]
@eddyleelin let's work closely with @jpkrohling to get the tooling aligned and built.

**Comment by Aneurysm9** [comment]
> I believe this to be the case, but can you please confirm that the regular Go tools will continue to be used for other purposes since go.mod is still the source of truth for them?

Yes, the proposal is to use additional tooling to simplify the management of tagging modules in a repo when not all modules should be tagged with the same versions.  Anything depending on those modules will do so using the ordinary Go toolchain.

> The intermediary branch will act as tags then, right? The branch `stable-v1` will be used by downstream dependencies of the collector when they resolve the transitive dependencies. Like:
> 
> 
> ```
> my-component -> groupbytrace v0.30.0 -> groupbysignal stable-v1
> ```

There will be no intermediary branches.  `stable-v1` is an example of a named module set managed by the tooling.  It would include a version declaration that would apply to all modules in the group by creating appropriate git tags.  The dependency graph you listed would be like so, presuming that `stable-v1` declared `v1.0.0` ans `groupbysignal` was contained in `stable-v1`:

```
my-component -> groupbytrace v0.30.0 -> groupbysignal v1.0.0
```

The `v1.0.0-RC1` and `v1.0.0-RC2` releases of stable components of the Go API/SDK have been tagged using this process, along with the `v0.21.0` and `v0.22.0` releases of experimental metrics components.

**Comment by jpkrohling** [comment]
> is an example of a named module set

Makes sense now. There cannot be interdependencies between modules within the module set, right?

**Comment by eddyleelin** [comment]
> There cannot be interdependencies between modules within the module set, right?

There actually can be interdependencies between modules within a module set (and that is often the case); for example, the `oteltest` Module does import `otel`, which are both in the `stable-v1` Module Set. As Anthony mentioned,

> stable-v1 is an example of a named module set managed by the tooling. It would include a version declaration that would apply to all modules in the group by creating appropriate git tags.

In other words, users can see that the entire "set of Modules" listed within the `stable-v1` Module Set are versioned together as a sort of "Multi-Module" grouping, which is basically an extension of how a single Module can group together multiple packages with the same version!

**Comment by jpkrohling** [comment]
> The v1.0.0-RC1 and v1.0.0-RC2 releases of stable components of the Go API/SDK have been tagged using this process

Looking at the commit for RC2 made it clear to me. For some reason, I thought we were trying to avoid `replaces` in the go.mod, but this is not the case. All in all, this is an interesting hack, as you are referencing a tag that doesn't exist at the time of the commit, but nobody is consuming it until the tag gets created.

I'm good with the proposal.

**Comment by bogdandrutu** [comment]
No longer applies to core, we are going to enable the tool in contrib.
